cvat/apps/engine/signals.py

```diff
@@ -30,18 +26,45 @@ def update_task_status(instance, **kwargs):
         db_task.status = status
         db_task.save()
 
-@receiver(post_save, sender=User, dispatch_uid="create_a_profile_on_create_a_user")
-def create_profile(instance, **kwargs):
+@receiver(post_save, sender=User,
+    dispatch_uid=__name__ + ".create_profile_handler")
+def __create_profile_handler(instance, **kwargs):
     if not hasattr(instance, 'profile'):
         profile = Profile()
         profile.user = instance
         profile.save()
 
-@receiver(post_delete, sender=Task, dispatch_uid="delete_task_files_on_delete_task")
-def delete_task_files_on_delete_task(instance, **kwargs):
+@receiver(post_delete, sender=Project,
+    dispatch_uid=__name__ + ".delete_project_handler")
+def __delete_project_handler(instance, **kwargs):
+    shutil.rmtree(instance.get_dirname(), ignore_errors=True)
+
+@receiver(post_delete, sender=Task,
+    dispatch_uid=__name__ + ".delete_task_handler")
+def __delete_task_handler(instance, **kwargs):
     shutil.rmtree(instance.get_dirname(), ignore_errors=True)
+    if instance.data and not instance.data.tasks.exists():
+        instance.data.delete()
 
+    try:
+        if instance.project: # update project
+            db_project = instance.project
+            db_project.save()
+    except Project.DoesNotExist:
+        pass # probably the project has been deleted
 
-@receiver(post_delete, sender=Data, dispatch_uid="delete_data_files_on_delete_data")
-def delete_data_files_on_delete_data(instance, **kwargs):
+@receiver(post_delete, sender=Job,
+    dispatch_uid=__name__ + ".delete_job_handler")
+def __delete_job_handler(instance, **kwargs):
+    shutil.rmtree(instance.get_dirname(), ignore_errors=True)
```

Comment:
I think `ignore_errors=True` is a bad idea, since that way we won't learn about any problems with deleting the directory. It would be better to pass an `onerror` callback that would log the error.

Reply:
@SpecLad , in general I agree with the proposal. But our logging system right now depends on the DB objects and files in deleted directories. Such change can lead to unexpected problems. Let's wait while we have a new logging system from Andrey. It should write events directly into ClickHouse DB.

Reply:
Okay, but can you at least add a TODO explaning this?

Reply:
I have added TODO comment